x-pack/metricbeat/module/azure/monitor/client.go

```diff
@@ -0,0 +1,238 @@
+	values       []MetricValue
+	timeGrain    string
+}
+
+// Dimension represents the azure metric dimension details
+type Dimension struct {
+	name  string
+	value string
+}
+
+// MetricValue represents the azure metric values
+type MetricValue struct {
+	name      string
+	average   *float64
+	min       *float64
+	max       *float64
+	total     *float64
+	count     *float64
+	timestamp time.Time
+}
+
+// NewClient instantiates the an Azure monitoring client
+func NewClient(config azure.Config) (*Client, error) {
+	azureMonitorService, err := Init(config.ClientID, config.ClientSecret, config.TenantID, config.SubscriptionID)
+	if err != nil {
+		return nil, err
+	}
+	client := &Client{
+		azureMonitorService: azureMonitorService,
+		config:              config,
+	}
+	client.resources.refreshInterval = config.RefreshListInterval
+	return client, nil
+}
+
+// InitResources returns the list of resources and maps them.
+func (client *Client) InitResources() error {
+	if !client.resources.expired() {
+		return nil
+	}
+	var metrics []Metric
+	if len(client.config.Resources) == 0 {
+		return errors.New("no resource options were configured")
+	}
+	for _, resource := range client.config.Resources {
+		//check all configuration options in order to identify resources : resourceid, resource group and resource query
+		if resource.ID != "" {
+			for _, metric := range resource.Metrics {
+				re, err := client.azureMonitorService.GetResourceById(resource.ID)
+				if err == nil {
+					met, err := client.mapMetric(metric, re)
+					if err != nil {
+						return err
+					}
+					metrics = append(metrics, met)
+				}
+			}
+		}
+		if resource.Group != "" {
+			resourceList, err := client.azureMonitorService.GetResourcesByResourceGroup(resource.Group, fmt.Sprintf("resourceType eq '%s'", resource.Type))
+			if err == nil {
+				for _, res := range resourceList {
+					for _, metric := range resource.Metrics {
+						met, err := client.mapMetric(metric, res)
+						if err != nil {
+							return err
+						}
+						metrics = append(metrics, met)
+					}
+				}
+			}
+
+		}
+		if resource.Query != "" {
+			resourceList, err := client.azureMonitorService.GetResourcesByResourceQuery(resource.Query)
+			if err == nil {
+				for _, res := range resourceList {
+					for _, metric := range resource.Metrics {
+						met, err := client.mapMetric(metric, res)
+						if err != nil {
+							return err
+						}
+						metrics = append(metrics, met)
+					}
+				}
+			}
+		}
+	}
+	client.resources.metrics = metrics
+	return nil
+}
+
+// GetMetricValues returns the specified metric data points for the specified resource ID/namespace.
+func (client *Client) GetMetricValues() error {
+	for i, metric := range client.resources.metrics {
+		client.resources.metrics[i].values = nil
+		endTime := time.Now().UTC()
+		startTime := endTime.Add(client.config.Period * (-1))
+		timespan := fmt.Sprintf("%s/%s", startTime.Format(time.RFC3339), endTime.Format(time.RFC3339))
```

Comment:
Are the metrics that are returned for this timespan already calculated metrics? For example if 2 metrics happened inside this timespan, will we get 1 or 2 results for 1 metric?